### DISCUSSION TITLE
Works once...then not again? And jumps me to other parts of my code?

### DISCUSSION THREAD

**Discussion by dave-linsalata** [photostructure/exiftool-vendored.js#107] [Q&A] [answered]
New to javascript and psyched to use this! However, I'm running into two problems I'd love some advice on:

1. When I transplant a working version of your modified usage code into my program, it works once...kind of: it can print the result to the console, but then hits the issue I list next. However, subsequent runs don't even get the console printing working. 
2. I also seem to be getting silent fails of some sort, such that the code returns and doesn't trip my "catch" statement. I've been trying to trace the execution, but haven't gotten anywhere. In fact, the return seems to jump the execution to a much later part of my code somehow, and I have no idea why.

I'm sure I'm missing something obvious in how this works, and I'm still getting comfortable with promises/async/await, but I can't for the life of me figure out why this doesn't work. Thanks!

_More details_

I can use exiftool to pull the information for a.mov file using either your first example's format... 

```
const exiftool = require("exiftool-vendored").exHiftool
exiftool
    .read('./images/2021-12-04 21.58.26.mov')
    .then((read) => {
        console.log(read);
    })
    .then(() => exiftool.end());
```
...or your second's...
```
const ExifTool = require("exiftool-vendored").ExifTool
const exiftool = new ExifTool({ taskTimeoutMillis: 5000 })
async function readFile (filename)
{
    const data = await exiftool.read(filename); 
    exiftool.end();
    return data;
}
readFile ('./images/2021-12-04 21.58.26.mov')
    .then (res => {
        console.log(res);
        }) 
```

However, I can't get it to work when I move it over to my code, neither using the first method...

```
  if (hasGPSdataPhotoList[i].fileName.includes('.mov')) {
            try {
                readFile (path.join("./images/", hasGPSdataPhotoList[i].fileName))
                     .then (result => {
                         console.log(result);
                         }) 
                }
                catch (error) {
                    console.log("Error reading Exif file data for: " + hasGPSdataPhotoList[i].fileName);
                    hasGPSdataPhotoList[i].hasGPSData = false;
                    originalPhotoList[i].hasGPSData = false;
                }
        }

// ... other code here ... 

async function readFile (filename) {
     const data = await exiftool.read(filename); 
     return data;
 }
```

...nor the second...

```
        if (hasGPSdataPhotoList[i].fileName.includes('.mov')) {
            try {
                exiftool
                    .read(path.join("./images/", hasGPSdataPhotoList[i].fileName))
                    .then((read) => {
                        console.log(read);
                        console.log(read.GPSLatitude);
                    })
                    .then(() => exiftool.end());
                catch (error) {
                    console.log("Error reading Exif file data for: " + hasGPSdataPhotoList[i].fileName);
                    hasGPSdataPhotoList[i].hasGPSData = false;
                    originalPhotoList[i].hasGPSData = false;
                }

```

**Comment by mceachen** [comment, marked as answer]
If nothing is being output to console, I suspect the main loop is not awaiting your promise chain, so node thinks it can ignore it and exit.

Both of the following examples work as expected on Node 16.13.1:

A more readable async/await style:

```js
const exiftool = require("exiftool-vendored").exiftool;

async function run() {
  const tags = await exiftool.read("../../examples/tiny.jpg");
  console.log(tags);
  await exiftool.end();
  console.log("finished.")
}

void run();
```

And promise .then chains:

```js
const exiftool = require("exiftool-vendored").exiftool;

void exiftool
  .read("../../examples/tiny.jpg")
  .then(console.log)
  .then(() => exiftool.end())
  .then(() => console.log("finished!"));
```

There are _tons_ of promise tutorials online, BTW. Please check at least a couple of them out.

**Reply by dave-linsalata** [reply]
Thanks for the response! Good point - I'll keep digging into the promises element. 

And thanks for the tip re: the tutorials: I've read through a number of guides (e.g. the developer.mozilla.org ones), but I'll keep looking for more to flesh out the knowledge.
